File: zig/src/candlestick_patterns/patterns/long_legged_doji.zig
/// Long Legged Doji: a one-candle pattern.
///
/// Must have:
/// - doji body (very small real body),
/// - one or both shadows are long.
///
/// Returns:
///     Continuous float in [0, 100].  Higher = stronger signal.

const cp = @import("../candlestick_patterns.zig");
const operators = @import("fuzzy").operators;

const CandlestickPatterns = cp.CandlestickPatterns;
const CriterionState = cp.CriterionState;
const lowerShadow = cp.lowerShadow;
const realBodyLen = cp.realBodyLen;
const upperShadow = cp.upperShadow;

pub fn longLeggedDoji(self: *const CandlestickPatterns) f64 {
            if (!self.enough(1, &[_]*const CriterionState{&self.doji_body, &self.long_shadow})) return 0.0;

    const b = self.bar(1);
    const mu_doji = self.muLessCs(realBodyLen(b.o, b.c), &self.doji_body, 1);
    const mu_long_us = self.muGreaterCs(upperShadow(b.o, b.h, b.c), &self.long_shadow, 1);
    const mu_long_ls = self.muGreaterCs(lowerShadow(b.o, b.l, b.c), &self.long_shadow, 1);
    const mu_any_long = operators.sMax(mu_long_us, mu_long_ls);

    const confidence = operators.tProductAll(&.{mu_doji, mu_any_long});
    return confidence * 100.0;
}

File: ts/candlestick-patterns/patterns/long-legged-doji.ts
/** Long Legged Doji pattern (1-candle). */
import { CandlestickPatternsEngine } from '../core/engine.ts';
import { tProductAll, sMax } from '../../fuzzy/index.ts';
import { realBody, upperShadow, lowerShadow } from '../core/primitives.ts';

/**
 * Long Legged Doji: a one-candle pattern.
 *
 * Must have:
 * - doji body (very small real body),
 * - one or both shadows are long.
 *
 * Returns:
 *     Continuous float in [0, 100].  Higher = stronger signal.
 */
export function longLeggedDoji(cp: CandlestickPatternsEngine): number {
    if (!cp.enough(1, cp.dojiBody, cp.longShadow)) return 0.0;

    const b = cp.bar(1);
    const muDoji = cp.muLessCS(realBody(b.o, b.c), cp.dojiBody, 1);
    const muLongUS = cp.muGreaterCS(upperShadow(b.o, b.h, b.c), cp.longShadow, 1);
    const muLongLS = cp.muGreaterCS(lowerShadow(b.o, b.l, b.c), cp.longShadow, 1);
    const muAnyLong = sMax(muLongUS, muLongLS);

    return tProductAll(muDoji, muAnyLong) * 100.0;
}
